### PR TITLE
Add more optimizations in the graph construction pipeline

### PR DESCRIPTION
**Pull request by zheng-da**
*Description of changes:*
This PR adds small optimizations and handles many errors in the input data properly.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/construct_graph.py [hunk 1]
```diff
@@ -838,4 +950,6 @@ def process_graph(args):
     argparser.add_argument("--num_partitions", type=int, default=1,
                            help="The number of graph partitions. " + \
                                    "This is only valid if the output format is DistDGL.")
+    argparser.add_argument("--skip_nonexist_edges", action='store_true',
```
Comment (thvasilo) [inline python/graphstorm/gconstruct/construct_graph.py:953]: Which are the cases where this happens? May be better to raise an error, depends on what we expect the user intent to be.

File: tests/unit-tests/test_construct_graph.py [hunk 2]
```diff
@@ -116,7 +117,102 @@ def test_label():
     assert np.sum(res['val_mask']) == 1
     assert np.sum(res['test_mask']) == 1
 
+def test_id_map():
```
Comment (thvasilo) [inline tests/unit-tests/test_construct_graph.py:120]: Might be better to break out individual tests instead of testing all functionality in one test unit.

File: tests/unit-tests/test_construct_graph.py [hunk 3]
```diff
@@ -116,7 +117,102 @@ def test_label():
...
+    str_ids = np.array([i for i in range(10)])
+    id_map = IdMap(str_ids)
+    try:
+        rand_ids = np.array([str(random.randint(0, len(str_ids))) for _ in range(5)])
+        remap_ids, idx = id_map(rand_ids)
+        raise ValueError("fails")
+    except:
+        pass
+
+def test_map_node_ids():
```
Comment (thvasilo) [inline tests/unit-tests/test_construct_graph.py:161]: Same here, unit test maintenance is much easier when they are testing a single-input single-output scenario.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 4]
```diff
@@ -409,35 +465,82 @@ def parse_edge_data(in_file, feat_ops, src_id_col, dst_id_col, edge_type,
...
-    dict : the key is the original ID and the value is the new ID.
     """
-    return {id1: i for i, id1 in enumerate(ids)}
+    def __init__(self, ids):
+        self._ids = {id1: i for i, id1 in enumerate(ids)}
+
+    def __len__(self):
+        return len(self._ids)
+
+    def __call__(self, ids):
```
Comment (thvasilo) [inline python/graphstorm/gconstruct/construct_graph.py:512]: Not clear what the expected behavior is here, better to add a docstring. Why not make this into a function with a descriptive name instead of using `__call__`?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 5]
```diff
@@ -370,8 +372,60 @@ def parse_node_data(in_file, feat_ops, node_id_col, label_conf,
...
+        if skip_nonexist_edges:
+            print(f"dest nodes of {src_type} do not exist: {dst_ids[bool_mask]}")
+        else:
+            raise ValueError(f"dest nodes of {src_type} do not exist: {dst_ids[bool_mask]}")
+        # We need to remove the source nodes as well.
+        src_ids = src_ids[idx]
+    dst_ids = new_dst_ids
+    return src_ids, dst_ids
+
 def parse_edge_data(in_file, feat_ops, src_id_col, dst_id_col, edge_type,
```
Comment (thvasilo) [inline python/graphstorm/gconstruct/construct_graph.py:427]: This function now has 9 arguments. Refactoring this will be difficult. Consider creating a wrapper configuration object that groups together relevant arguments (using a `dataclass`) or breaking into multiple functions.

A good rule of thumb (that [pylint](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-arguments.html) uses) is for function to not have more than 5 arguments if possible.

A good way to avoid having to pass state around like this is to wrap functions like these in a class that you can initialize with all the shared state.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 6]
```diff
@@ -166,8 +166,10 @@ def __call__(self, strs):
             t = self.tokenizer(s, max_length=self.max_seq_length,
                                truncation=True, padding='max_length', return_tensors='pt')
             tokens.append(t['input_ids'])
-            att_masks.append(t['attention_mask'])
-            type_ids.append(t['token_type_ids'])
+            # The masks are small integers. We don't need to use int4 or int8 to store them.
+            # This can signficantly reduce memory consumption.
```
Comment (thvasilo) [inline python/graphstorm/gconstruct/construct_graph.py:170]: Did we measure the memory consumed here? How much are we saving by using these representations? (vs. increasing the chances we run into overflow errors)

File: python/graphstorm/gconstruct/construct_graph.py [hunk 7]
```diff
@@ -166,8 +166,10 @@ def __call__(self, strs):
             t = self.tokenizer(s, max_length=self.max_seq_length,
                                truncation=True, padding='max_length', return_tensors='pt')
             tokens.append(t['input_ids'])
-            att_masks.append(t['attention_mask'])
-            type_ids.append(t['token_type_ids'])
+            # The masks are small integers. We don't need to use int4 or int8 to store them.
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:169]: `We can use int4 or int8` instead of `We don't need to use int4 or int8` ?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 8]
```diff
@@ -370,8 +372,60 @@ def parse_node_data(in_file, feat_ops, node_id_col, label_conf,
             feat_data[key] = val
     return (data[node_id_col], feat_data)
 
+def map_node_ids(src_ids, dst_ids, edge_type, node_id_map, skip_nonexist_edges):
+    """ Map node IDs of source and destination nodes of edges.
+
+    In the ID mapping, we need to handle many errors in the input data. For example,
+    we need to handle the case that endpoint nodes of edges don't exist; we also need
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:379]: Can we directly list how many different errors we handle in this function?
`For example` is vague.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 9]
```diff
@@ -370,8 +372,60 @@ def parse_node_data(in_file, feat_ops, node_id_col, label_conf,
...
+        The key is the node type and value is IdMap or IdentityMap.
+    skip_nonexist_edges : bool
+        Whether or not to skip edges whose endpoint nodes don't exist.
+
+    Returns
+    -------
+    tuple of tensors : the remapped source and destination node IDs.
+    """
+    src_type, _, dst_type = edge_type
+    new_src_ids, idx = node_id_map[src_type](src_ids)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:401]: Does `idx` represent the nodes that exists in node_id_map[src_type] ? Can we give it a better name?

File: python/graphstorm/gconstruct/construct_graph.py (hunk 4 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/construct_graph.py:512]: Agree with Theo. A function would be better.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 10]
```diff
@@ -409,35 +465,82 @@ def parse_edge_data(in_file, feat_ops, src_id_col, dst_id_col, edge_type,
...
+                assert np.issubdtype(ids.dtype, np.integer), \
+                        "The key of ID map is integer, input IDs should also be integers. " \
+                        + f"But get {type(ids[0])}."
+            else:
+                raise ValueError(f"Unsupported key data type: {type(id_)}")
+            break
+
+        new_ids = []
+        idx = []
+        for i, id_ in enumerate(ids):
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:530]: Can you add some docstring describing what you are doing?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 11]
```diff
@@ -755,6 +866,15 @@ def process_edge_data(process_confs, node_id_map, num_processes):
 
     return edges, edge_data
 
+def verify_confs(confs):
+    """ Verify the configuration of the input data.
+    """
+    ntypes = {conf['node_type'] for conf in confs["node"]}
+    etypes = [conf['relation'] for conf in confs["edge"]]
+    for src_type, _, dst_type in etypes:
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:874]: What if a node type does not have any node feature? Are we still ask users to provide node files for that node type?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 12]
```diff
@@ -802,18 +909,23 @@ def process_graph(args):
...
     for ntype in node_id_map:
-        map_data = {}
-        map_data["orig"] = np.array(list(node_id_map[ntype].keys()))
-        map_data["new"] = np.array(list(node_id_map[ntype].values()))
-        write_data_parquet(map_data, os.path.join(args.output_dir, ntype + "_id_remap.parquet"))
+        kv_pairs = node_id_map[ntype].get_key_vals()
+        if kv_pairs is not None:
+            map_data = {}
+            map_data["orig"], map_data["new"] = kv_pairs
+            write_data_parquet(map_data, os.path.join(args.output_dir,
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:927]: Can user choose the data format?
Or can we use the same format as input data?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 13]
```diff
@@ -802,18 +909,23 @@ def process_graph(args):
...
+                                                      ntype + "_id_remap.parquet"))
 
 if __name__ == '__main__':
     argparser = argparse.ArgumentParser("Preprocess graphs")
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:931]: We should also make a doc describing all the input arguments of construct_graph.

File: tests/unit-tests/test_construct_graph.py (hunk 2 above)
Comment (classicsong) [inline tests/unit-tests/test_construct_graph.py:120]: Please add a description for this test.

File: tests/unit-tests/test_construct_graph.py [hunk 14]
```diff
@@ -116,7 +117,102 @@ def test_label():
...
+    # Test the case that destination node IDs don't exist.
+    src_ids = np.array([str(random.randint(0, len(str_src_ids) - 1)) for _ in range(15)])
+    dst_ids = np.array([str(random.randint(0, 20)) for _ in range(15)])
+    try:
+        new_src_ids, new_dst_ids = map_node_ids(src_ids, dst_ids, ("src", None, "dst"),
+                                                id_map, False)
+        raise ValueError("fail")
+    except:
+        pass
+    # Test the case that destination node IDs don't exist and we skip non exist edges.
```
Comment (classicsong) [inline tests/unit-tests/test_construct_graph.py:206]: Shall we split it into multiple small unitests?

File: python/graphstorm/gconstruct/construct_graph.py (hunk 1 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:953]: some data (e.g., MAG graph) has this problem that some of the end point nodes don't exist in the node files. this option is to tell the pipeline to ignore these edges. by default, it raises an error. however, i would like to give users the option of skipping these edges. otherwise, the users have to deal with this problem themselves.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 5 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:427]: i try to understand the problem of this. is it about the number of arguments or the amount of work done in this function?
if we wrap multiple arguments into an object, it doesn't change anything except the number of arguments.

we cannot break this function because this is a entry function that runs in process pool.

Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:427]: i guess the easiest solution is to pass `process_conf` to the function directly.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 6 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:170]: a lot. Memory consumption reduces by almost 3 times.

Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:170]: attention mask is bool. token type ids are small numbers according to the documentation of huggingface.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 9 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:401]: node location in `src_ids`

File: python/graphstorm/gconstruct/construct_graph.py (hunk 11 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:874]: yes, we currently require users to provide node files for every node type. computing unique node IDs is very expensive operation. it requires appropriate optimization. I don't think I'll support this in v0.1. I want to see users' feedbacks first.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 12 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:927]: not sure what is the best option here. maybe we should allow users to provide an argument to specify the format. or maybe JSON is the best format here.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 13 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:931]: yes, just like training/inference script.

File: tests/unit-tests/test_construct_graph.py (hunk 14 above)
Comment (zheng-da) [reply tests/unit-tests/test_construct_graph.py:206]: do you mean split it into multiple `check_xxx` functions?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 15]
```diff
@@ -409,35 +462,109 @@ def parse_edge_data(in_file, feat_ops, src_id_col, dst_id_col, edge_type,
...
+        Returns
+        -------
+        tuple of tensors : the tensor of new IDs, the location of the IDs in the input ID tensor.
+        """
+        return ids, np.arange(len(ids))
+
+    def get_key_vals(self):
+        """ Get the key value pairs.
+        """
+        return None
```
Comment (thvasilo) [inline python/graphstorm/gconstruct/construct_graph.py:504]: What is the purpose of having this function if it just returns `None`? If we want to have the same exact interface for "IdentityMap" and "IdMap" they should probably belong to a class hierarchy with a common interface.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 16]
```diff
@@ -409,35 +462,109 @@ def parse_edge_data(in_file, feat_ops, src_id_col, dst_id_col, edge_type,
...
+                assert np.issubdtype(ids.dtype, np.integer), \
+                        "The key of ID map is integer, input IDs should also be integers. " \
+                        + f"But get {type(ids[0])}."
+            else:
+                raise ValueError(f"Unsupported key data type: {type(id_)}")
+            break
+
+        # If the input ID exists in the ID map, map it to a new ID
+        # and keep its location in the input ID array.
+        # otherwise, skip the ID.
```
Comment (thvasilo) [inline python/graphstorm/gconstruct/construct_graph.py:551]:
```suggestion
        # Otherwise, skip the ID.
```

**Review by thvasilo — COMMENTED** [review]
A couple of minor changes requested, the logic seems fine.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 17]
```diff
@@ -755,6 +884,15 @@ def process_edge_data(process_confs, node_id_map, num_processes):
 
     return edges, edge_data
 
+def verify_confs(confs):
+    """ Verify the configuration of the input data.
+    """
+    ntypes = {conf['node_type'] for conf in confs["node"]}
+    etypes = [conf['relation'] for conf in confs["edge"]]
+    for src_type, _, dst_type in etypes:
+        assert src_type in ntypes, f"source node type {src_type} does not exist."
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:893]: Can we also tell user how to fix the problem or what is the root cause of the problem? For example, they do not add src_node_type in the config.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 12 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/construct_graph.py:945]: I agree.

**Review by classicsong — APPROVED** [review]
LGTM.
But please resolve Theo's comments.
